File: car_rental_system/car_rental_system/late_approval_template/late_approval_template.py
# Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
# For license information, please see license.txt

import datetime
import json

import frappe
from frappe.model.document import Document
from frappe.utils import getdate


class LateApprovalTemplate(Document):
	pass

@frappe.whitelist()
def get_employees(
	date: str | datetime.date,
	department: str | None = None,
	branch: str | None = None,
	company: str | None = None,
	shift:str | None = None,
) -> dict[str, list]:
	#filters = {"status": "Active", "date_of_joining": ["<=", date]}
	filters={}
	for field, value in {"department": department, "attendance_date": date, "company": company, "shift":shift,"docstatus": 1, "late_entry":1,}.items():
		if value:
			filters[field] = value

	# employee_list = frappe.get_list(
	# 	"Employee", fields=["employee", "employee_name"], filters=filters, order_by="employee_name"
	# )
	attendance_list = frappe.get_list(
		"Attendance",
		fields=["employee", "employee_name", "status"],
		filters=filters,
		# filters={
		# 	"attendance_date": date,
		# 	"docstatus": 1,
		# 	"late_entry":1,
		# 	"shift": shift,
		# },
		order_by="employee_name",
	)

	# get_attendance = _get_attendance(employee_list, attendance_list)

	return {"attendance": attendance_list}
	#return {"marked": attendance_list, "unmarked": unmarked_attendance}

# def _get_attendance(employee_list: list[dict], attendance_list: list[dict]) -> list[dict]:
# 	marked_employees = [entry.employee for entry in attendance_list]
# 	get_attendance = []

# 	for entry in employee_list:
# 		if entry.employee not in marked_employees:
# 			get_attendance.append(entry)

# 	return get_attendance

@frappe.whitelist()
def mark_employee_late(
	employee_list: list | str,
	date: str | datetime.date,
	late_approved: int | None = None,
) -> None:
	if isinstance(employee_list, str):
		employee_list = json.loads(employee_list)

	for employee in employee_list:
		
		# attendance = frappe.get_doc(
		# 	dict(
		# 		doctype="Attendance",
		# 		employee=employee,
		# 		attendance_date=getdate(date),
		# 		late_entry=late_approved
		# 	)
		# )
		doc = frappe.get_last_doc('Attendance',filters={"employee":employee,"attendance_date":getdate(date)})
		#doc = frappe.get_doc("Attendance",attendance)
		doc.late_entry = late_approved

		doc.db_update()
		# attendance.update()
		# attendance.submit()

File: car_rental_system/car_rental_system/late_approval_template/late_approval_template.js
// Copyright (c) 2024, Frappe Technologies Pvt. Ltd. and contributors
// For license information, please see license.txt

frappe.ui.form.on("Late Approval Template", {
	refresh(frm) {
		frm.trigger("reset_attendance_fields");
		frm.trigger("load_employees");
		frm.trigger("set_primary_action");
    },
    onload(frm) {
		frm.set_value("date", frappe.datetime.get_today());
	},

	date(frm) {
		frm.trigger("load_employees");
	},

	department(frm) {
		frm.trigger("load_employees");
	},

	branch(frm) {
		frm.trigger("load_employees");
	},

	company(frm) {
		frm.trigger("load_employees");
	},

	status(frm) {
		frm.trigger("set_primary_action");
	},

	reset_attendance_fields(frm) {
		frm.set_value("shift", "");
		frm.set_value("late_approved", 0);
	},
    load_employees(frm) {
		if (!frm.doc.date) return;

		frappe
			.call({
				method: "hrms.hr.doctype.late_approval_template.late_approval_template.get_employees",
				args: {
					date: frm.doc.date,
					department: frm.doc.department,
					branch: frm.doc.branch,
					company: frm.doc.company,
				},
			})
			.then((r) => {
				frm.employees = r.message["attendance"];

				if (r.message["attendance"].length > 0) {
					unhide_field("select_employees_section");
					//unhide_field("late_details_section");
					frm.events.show_unmarked_employees(frm, r.message["attendance"]);
				} else {
					hide_field("select_employees_section");
					//hide_field("late_details_section");
				}
			});
	},

	show_unmarked_employees(frm, attendance) {
        console.log(attendance);
		const $wrapper = frm.get_field("employees_html").$wrapper;
		$wrapper.empty();
		const employee_wrapper = $(`<div class="employee_wrapper">`).appendTo($wrapper);

		frm.employees_multicheck = frappe.ui.form.make_control({
			parent: employee_wrapper,
			df: {
				fieldname: "employees_multicheck",
				fieldtype: "MultiCheck",
				select_all: true,
				columns: 4,
				get_data: () => {
					return attendance.map((employee) => {
						return {
							label: `${employee.employee} : ${employee.employee_name}`,
							value: employee.employee,
							checked: 0,
						};
					});
				},
			},
			render_input: true,
		});

		frm.employees_multicheck.refresh_input();
	},
    set_primary_action(frm) {
		frm.disable_save();
		frm.page.set_primary_action(__("Mark Late Approval"), () => {
			if (frm.employees.length === 0) {
				frappe.msgprint({
					message: __(
						"Late Approval for all the employees under this criteria has been marked already.",
					),
					title: __("Late Approved"),
					indicator: "green",
				});
				return;
			}

			if (frm.employees_multicheck.get_checked_options().length === 0) {
				frappe.throw({
					message: __("Please select the employees you want to mark late approval for."),
					title: __("Mandatory"),
				});
			}

			// if (!frm.doc.shift) {
			// 	frappe.throw({
			// 		message: __("Please select the shift for employees."),
			// 		title: __("Mandatory"),
			// 	});
			// }

			frm.trigger("mark_late_approval");
		});
	},

	mark_late_approval(frm) {
		const marked_employees = frm.employees_multicheck.get_checked_options();

		frappe
			.call({
				method: "hrms.hr.doctype.late_approval_template.late_approval_template.mark_employee_late",
				args: {
					employee_list: marked_employees,
					date: frm.doc.date,
					late_entry: frm.doc.late_approved,
					shift: frm.doc.shift,
				},
				freeze: true,
				freeze_message: __("Marking Late Approvals"),
			})
			.then((r) => {
				if (!r.exc) {
					frappe.show_alert({
						message: __("Lates Approved successfully"),
						indicator: "green",
					});
					frm.refresh();
				}
			});
	},
});
